### PR TITLE
Add syntax tests for Yul objects

### PR DESCRIPTION
**Pull request by axic**
I did this before realising `ObjectParser.cpp` contains very similar test cases, but having them in this format I think is nicer. We should probably remove the duplicated cases from `ObjectParser.cpp` if we merge this.

### CODE REVIEW

**Comment by chriseth** [comment]
Why not use the object parser for regular yul syntax tests instead?

**Comment by chriseth** [comment]
It falls back to "code only" mode if the text does not start with `object`.

**Comment by axic** [comment]
> Why not use the object parser for regular yul syntax tests instead?

That was my first choice (added all these checks into `yulSyntaxTests`), but that posed some issues. I can't fully remember what but it seemed reasonable to be separate.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `3a083c29ca709407940d3f086057b2984b88e39f`:
```
Error: Trailing whitespace found:
 test/libyul/yulObjectSyntaxTests/conflict_data_data.yul:3: 
 test/libyul/yulObjectSyntaxTests/conflict_object_data.yul:3: 
 test/libyul/yulObjectSyntaxTests/conflict_object_object.yul:3: 

```
Please check that your changes are working as intended.

**Comment by axic** [comment]
> It falls back to "code only" mode if the text does not start with object.

It may do that but the object mode is just not working properly:
```
yulSyntaxTests/objects/multiple_root_object.yul: FAIL
  Contract:
    object "A" {
      code { }
    }
    object "B" {
      code { }
    }

  Expected result:
    ParserError 2314: (26-32): Expected end of source but got identifier
  Obtained result:
    ParserError 2314: (0-6): Expected '{' but got identifier
```

```
yulSyntaxTests/objects/conflict_object_object.yul: FAIL
  Contract:
    object "A" {
      code {}
    
      object "B" {
        code {}
      }
      object "B" {
        code {}
      }
    }

  Expected result:
    ParserError 8794: (66-69): Object name "B" already exists inside the containing object.
  Obtained result:
    ParserError 2314: (0-6): Expected '{' but got identifier
```

etc

**Comment by axic** [comment]
@chriseth I think you mean the `yulOptimiserTests` which supports both?

**Comment by axic** [comment]
Fixed `libyul/SyntaxTests` to parse objects too, better this way?

File: test/libyul/SyntaxTest.cpp [hunk 1]
```diff
@@ -26,6 +26,8 @@
 #include <test/libyul/Common.h>
 #include <test/libyul/SyntaxTest.h>
 
+#include <test/libsolidity/util/SoltestErrors.h>
```
Comment (axic) [inline test/libyul/SyntaxTest.cpp:29]: For `soltestAssert`.

**Comment by axic** [comment]
More tests can be moved and things can be dropped from `test/libyul/ObjectParser.cpp`, but would like to get some confirmation first that this direction is approved.

@chriseth what do you think?

File: test/libyul/yulSyntaxTests/objects/datacopy.yul [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+    datacopy(0, 1, 2)
+
+    // datacopy also accepts pretty much anything which can be turned into a number
```
Comment (ekpyron) [inline test/libyul/yulSyntaxTests/objects/datacopy.yul:4]: A bit of a weird comment :-). It just accepts arbitrary yul expressions as arguments... but fine!

**Review by ekpyron — DISMISSED** [review]
One minor comment about a comment... and I'd squash it to a single commit... but fine with merging!

**Review by ekpyron — DISMISSED** [review]

**Review by ekpyron — APPROVED** [review]

**Pull request merged**